Improve error handling

medikoo opened the pull request:
- Do not pass _handled_ errors via _unhandled rejections_ channel - it's a logic error, that's also  indirect cause of #6441 
- Use default error logging for _uncaught exceptions_, but ensure process exits immediately after report is made (improves UX)
- Propagate _unhandled rejections_ as _uncaught excpeptions_ but only if no other _unhandled rejection_ handlers were registered (it's to resemble eventual future Node.js behavior, and to allow users to opt out from this handling - related issue: https://github.com/dherault/serverless-offline/issues/255)
- Use default error logging for non error exceptions (improves UX)  
   _~~(Note: here to distinguish error instances I decided to rely on `type/error/is`, which makes it a new dependency. It's the best way known to me to detect a native error, and dependency on it's own is light (has no other deps), and has many other useful type validation utils which we may want to rely on. [It's also already used in `@serverless/enterprise-plugin`](https://github.com/serverless/enterprise-plugin/pull/191#discussion_r302853461))~~_
- Do not show `SLS_DEBUG=*` suggestion when it's already used
- Unconditionally expose stack trace for non user errors (so not `ServerlessError`'s, and others we recognize)  _(If they happen they're definitely a result of bug, and it'll be good to have stack trace provided in issue reports)_
- Improve variable naming

**_Is it a breaking change?:_** NO

pmuens commented on lib/classes/Error.js:
```diff
@@ -42,67 +58,69 @@ module.exports.ServerlessError = class ServerlessError extends Error {
+  consoleLog(chalk.yellow(`     Node Version:              ${nodeVersion}`));
+  consoleLog(chalk.yellow(`     Serverless Version:        ${slsVersion}`));
+  consoleLog(chalk.yellow(`     Enterprise Plugin Version: ${sfeVersion}`));
+  consoleLog(chalk.yellow(`     Platform SDK Version:      ${sdkVersion}`));
```
We just changed the name of those values via https://github.com/serverless/serverless/pull/6501 so this PR might need a rebase / update.

medikoo replied:
Actually it was not changed here (in master it's still `Enterprise Plugin Version`). I can rename it, still isn't `Plugin` name too vague?

pmuens replied:
Yes, most agreed that it's too vague. According to JIRA that's the whole idea (vague == more flexible)

pmuens commented on lib/classes/Error.js:
```diff
@@ -2,6 +2,7 @@
 
 const chalk = require('chalk');
 const { inspect } = require('util');
+const isError = require('type/error/is');
```
Why don't we use https://lodash.com/docs/4.17.15#isError here?

medikoo replied:
My bad, I didn't actually check if Lodash offers something like that.

I've updated it to use `_.isError` (by looking into internals it looks good, and I agree in context of this single case it's not justified to introduce new dependency).

Still I think at some point it might be good if we derive from lodash, so we rely more on native constructs, and imo `type` stands as nice complement of runtime type validation utils (which is not really covered by lodash) through which we may improve DX experience.

medikoo commented:
> I ran the code locally and apparently the recent changes introduce some hard to read error messages (see below). 

Good find, indeed it got worse for `YAMLException`.

First reason is that `YAMLException` is custom error constructed with ES5 means (no as class extension), so it was not detected as native error by `type/error/is`, and it was output through `inspect(error)`. 
I've improved error detection in `type` so ES5 custom errors are picked now, but I've also switched for now to `_.isError` which also recognizes it.

Other reason is that full stack trace was output as `YAMLException` was not recognized as _user error_. I've updated it so it's recognized now as such (like `ServerlessError` is)

medikoo commented:
> It looks like the only thing we need to update as well is the naming we've also changed via #6501.

Ok, updated
